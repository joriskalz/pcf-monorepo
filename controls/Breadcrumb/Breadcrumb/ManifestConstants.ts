export const enum ManifestPropertyNames {
    dataset = 'dataset',
}

export const enum ItemColumns {
    DisplayName = 'ItemDisplayName',
    Key = 'ItemKey',
    Clickable = 'ItemClickable',
}

export const enum InputEvents {
    SetFocus = 'SetFocus',
}

export const enum InputProperties {
    InputEvent = 'InputEvent',
    SelectedKey = 'SelectedKey',
}
