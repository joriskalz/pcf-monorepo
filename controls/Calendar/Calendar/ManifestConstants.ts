export const enum InputEvents {
    SetFocus = 'SetFocus',
}
