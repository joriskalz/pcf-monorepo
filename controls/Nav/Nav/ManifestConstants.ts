export const enum ManifestPropertyNames {
    dataset = 'dataset',
}

export const enum ItemColumns {
    DisplayName = 'ItemDisplayName',
    Key = 'ItemKey',
    Enabled = 'ItemEnabled',
    IconName = 'ItemIconName',
    IconColor = 'ItemIconColor',
    TextColor = 'ItemTextColor',
    ParentKey = 'ItemParentKey',
    Visible = 'ItemVisible',
    Expanded = 'ItemExpanded',
}

export const enum InputEvents {
    SetFocus = 'SetFocus',
}

export const enum InputProperties {
    SelectedKey = 'SelectedKey',
}
