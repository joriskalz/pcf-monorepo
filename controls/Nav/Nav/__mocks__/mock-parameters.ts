/* istanbul ignore file */

import { IInputs } from '../generated/ManifestTypes';
import { MockStringProperty, MockTwoOptionsProperty } from './mock-context';
import { MockDataSet } from './mock-datasets';

export function getMockParameters(): IInputs {
    return {
        AccessibilityLabel: new MockStringProperty(),
        CollapseByDefault: new MockTwoOptionsProperty(),
        InputEvent: new MockStringProperty(),
        SelectedKey: new MockStringProperty(),
        Theme: new MockStringProperty(),
        items: new MockDataSet([]),
    };
}
