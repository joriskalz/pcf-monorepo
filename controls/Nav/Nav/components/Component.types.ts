export interface NavItem {
    id: string;
    enabled?: boolean;
    name: string;
    iconName?: string;
    iconColor?: string;
    visible?: boolean;
    parentItemKey?: string;
    url: string;
    links: NavItem[];
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    data: any;
    key: string;
    isExpanded: boolean;
}

export interface CanvasNavProps {
    width?: number;
    height?: number;
    items: NavItem[];
    nestedItems: NavItem[];
    onSelected: (item: NavItem) => void;
    disabled?: boolean;
    tabIndex?: number;
    ariaLabel?: string;
    themeJSON?: string;
    selectedKey?: string;
    setFocus?: string;
    collapseByDefault?: boolean;
}
