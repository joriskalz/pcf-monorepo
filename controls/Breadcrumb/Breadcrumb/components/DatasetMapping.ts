import { IBreadcrumbItem } from '@fluentui/react';
import { ItemColumns } from '../ManifestConstants';
import { ICustomBreadcrumbItem } from './components.types';

export function getItemsFromDataset(dataset: ComponentFramework.PropertyTypes.DataSet): ICustomBreadcrumbItem[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return [getDummyAction('1'), getDummyAction('2'), getDummyAction('3')];
    }
    const keyIndex: Record<string, number> = {};
    return dataset.sortedRecordIds.map((id) => {
        const record = dataset.records[id];
        // Prevent duplicate keys by appending the duplicate index
        let key = record.getValue(ItemColumns.Key) as string;
        if (keyIndex[key] !== undefined) {
            keyIndex[key]++;
            key += `_${keyIndex[key]}`;
        } else keyIndex[key] = 1;
        return {
            id: record.getRecordId(),
            key: key,
            text: record.getFormattedValue(ItemColumns.DisplayName),
            clickable: record.getValue(ItemColumns.Clickable) as boolean,
            data: record,
        } as ICustomBreadcrumbItem;
    });
}

export function getBreadcrumbItems(
    items: ICustomBreadcrumbItem[],
    onClick: (ev?: React.MouseEvent<HTMLElement>, item?: ICustomBreadcrumbItem) => void,
): IBreadcrumbItem[] {
    return items.map((item) =>
        item.clickable === false
            ? {
                  id: item.id,
                  key: item.key,
                  text: item.text,
              }
            : {
                  id: item.id,
                  key: item.key,
                  text: item.text,
                  onClick: onClick,
              },
    ) as IBreadcrumbItem[];
}

function getDummyAction(key: string): ICustomBreadcrumbItem {
    return {
        id: key,
        key: key,
        text: 'Item ' + key,
        clickable: true,
    } as ICustomBreadcrumbItem;
}
