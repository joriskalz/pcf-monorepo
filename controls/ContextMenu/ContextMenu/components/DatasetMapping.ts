import { CanvasContextMenuItem } from './Component.types';
import {
    ContextualMenuItemType,
    getColorFromString,
    getShade,
    IButtonStyles,
    IContextualMenuItem,
    IIconProps,
    Shade,
} from '@fluentui/react';
import { ItemColumns } from '../ManifestConstants';
const CSS_IMPORTANT = ' !important';
export function getMenuItemsFromDataset(dataset: ComponentFramework.PropertyTypes.DataSet): CanvasContextMenuItem[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return [getDummyAction('1'), getDummyAction('2'), getDummyAction('3')];
    }
    const keyIndex: Record<string, number> = {};
    return dataset.sortedRecordIds.map((id) => {
        const record = dataset.records[id];
        // Prevent duplicate keys by appending the duplicate index
        let key = (record.getFormattedValue(ItemColumns.Key) as string) ?? '';
        if (keyIndex[key] !== undefined) {
            keyIndex[key]++;
            key += `_${keyIndex[key]}`;
        } else keyIndex[key] = 1;
        return {
            id: record.getRecordId(),
            key: key,
            name: record.getFormattedValue(ItemColumns.DisplayName),
            enabled: (record.getValue(ItemColumns.Enabled) as boolean) ?? undefined,
            iconName: record.getFormattedValue(ItemColumns.IconName),
            iconColor: record.getFormattedValue(ItemColumns.IconColor),
            parentItemKey: record.getFormattedValue(ItemColumns.ParentKey),
            iconOnly: (record.getValue(ItemColumns.IconOnly) as boolean) ?? undefined,
            checked: (record.getValue(ItemColumns.Checked) as boolean) ?? undefined,
            visible: (record.getValue(ItemColumns.Visible) as boolean) ?? undefined,
            isHeader: (record.getValue(ItemColumns.ItemHeader) as boolean) ?? undefined,
            topDivider: (record.getValue(ItemColumns.ItemTopDivider) as boolean) ?? undefined,
            divider: (record.getValue(ItemColumns.ItemDivider) as boolean) ?? undefined,
            data: record,
        } as CanvasContextMenuItem;
    });
}

function getDummyAction(key: string): CanvasContextMenuItem {
    return {
        id: key,
        key: key,
        name: 'Item ' + key,
        iconName: 'Unknown',
    } as CanvasContextMenuItem;
}

export function getCommandsWithChildren(
    items: CanvasContextMenuItem[],
    disabled: boolean,
    onClick: (ev?: React.MouseEvent<HTMLButtonElement>, item?: IContextualMenuItem | undefined) => boolean,
): IContextualMenuItem[] {
    return items
        .filter((i) => !i.parentItemKey && i.visible !== false)
        .map((i) => {
            return getCommandBarItemProps(items, i, disabled, onClick);
        });
}

export function getCommandBarItemProps(
    items: CanvasContextMenuItem[],
    item: CanvasContextMenuItem,
    disabled: boolean,
    onClick: (ev?: React.MouseEvent<HTMLButtonElement>, item?: IContextualMenuItem | undefined) => boolean,
): IContextualMenuItem {
    const subItems = items
        .filter((i) => i.parentItemKey === item.key && i !== item && i.visible !== false)
        .map((i) => {
            return getCommandBarItemProps(items, i, disabled, onClick);
        });
    const buttonStyles = {
        root: {
            background: 'rgba(255, 255, 255,0)',
        },
    } as IButtonStyles;

    const iconProps =
        item.iconName &&
        ({
            iconName: item.iconName,
        } as IIconProps);

    // If custom icon color, find the shade slots for the hover/pressed state
    if (iconProps && item.iconColor) {
        const iconColor = item.iconColor;
        iconProps.styles = { root: { color: item.iconColor + CSS_IMPORTANT } };

        // Attempt to set a darker shade for hover/pressed colors
        const primaryColor = getColorFromString(iconColor);
        if (primaryColor) {
            const iconColorDarker = getShade(primaryColor, Shade.Shade8);
            const secondaryColorStyle = { color: iconColorDarker?.str + CSS_IMPORTANT };
            buttonStyles.iconHovered = secondaryColorStyle;
            buttonStyles.iconPressed = secondaryColorStyle;
            buttonStyles.iconExpanded = secondaryColorStyle;
            buttonStyles.iconChecked = secondaryColorStyle;
            buttonStyles.iconExpandedHovered = secondaryColorStyle;
        }
    }

    const props = {
        key: item.key,
        text: item.name,
        disabled: item.enabled === false || disabled,
        onClick: onClick,
        iconOnly: item.iconOnly === true,
        iconProps: iconProps,
        canCheck: item.checked !== undefined,
        checked: item.checked,
        data: item,
        buttonStyles: buttonStyles,
    } as IContextualMenuItem;

    if (item.isHeader === true && subItems.length > 0) {
        // Semantic Section
        props.itemType = ContextualMenuItemType.Section;
        props.sectionProps = {
            title: item.name,
            topDivider: item.topDivider,
            bottomDivider: item.divider,
            items: subItems,
        };
    } else if (item.isHeader === true) {
        // Simple Section Break
        props.itemType = ContextualMenuItemType.Header;
    } else if (item.divider === true) {
        // Divider Break
        props.itemType = ContextualMenuItemType.Divider;
    } else {
        props.itemType = ContextualMenuItemType.Normal;
        props.subMenuProps =
            subItems.length > 0
                ? {
                      items: subItems,
                  }
                : undefined;
    }
    return props;
}
