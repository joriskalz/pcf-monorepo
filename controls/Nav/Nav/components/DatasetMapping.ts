import { NavItem } from './Component.types';
import { INavLink } from '@fluentui/react';
import { ItemColumns } from '../ManifestConstants';

export function getMenuItemsFromDataset(dataset: ComponentFramework.PropertyTypes.DataSet): NavItem[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return getDummyAction();
    }
    return dataset.sortedRecordIds.map((id) => {
        const record = dataset.records[id];
        return {
            id: record.getRecordId(),
            key: record.getValue(ItemColumns.Key) as string,
            name: record.getValue(ItemColumns.DisplayName) as string,
            enabled: record.getValue(ItemColumns.Enabled) as boolean,
            iconName: record.getValue(ItemColumns.IconName) as string,
            iconColor: record.getValue(ItemColumns.IconColor) as string,
            parentItemKey: record.getValue(ItemColumns.ParentKey) as string,
            visible: record.getValue(ItemColumns.Visible) as boolean,
            isExpanded: record.getValue(ItemColumns.Expanded) as boolean,
            data: record,
        } as NavItem;
    });
}

/**
 * This method is primarily to structure the links as a tree to support Fluent UI Nav
 * @param items
 * @returns
 */
export function getTreeBasedGrpLinks(items: NavItem[]): NavItem[] {
    let root: NavItem[] = [];
    //Creating a custom rootID for NavGroupLink root node.
    const rootID = 'rootID'.concat(items[0].key, items[items.length - 1].key);
    items.map((item) => {
        // identify root and orphan items and replace their parentkey with rootID
        if (item.parentItemKey !== undefined) {
            if (
                items
                    .map((item) => {
                        return item.key;
                    })
                    .indexOf(item.parentItemKey) === -1
            ) {
                item.parentItemKey = rootID;
            }
        } else {
            item.parentItemKey = rootID;
        }
    });
    //add custom root to accomodate orphan items or links
    items.unshift({ id: rootID, name: rootID, key: rootID, parentItemKey: null } as unknown as NavItem);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const keyMapping = items.reduce((prevGrp: any, currentGrp, i) => {
        prevGrp[currentGrp.key] = i;
        return prevGrp;
    }, {});
    items.map((grp) => {
        // Handle the root element
        if (grp.parentItemKey == null) {
            root = [grp] as unknown as NavItem[];
            return;
        }
        if (items[keyMapping[grp.key]].visible !== false) {
            const parentGrp = items[keyMapping[grp.parentItemKey]] as unknown as INavLink;
            parentGrp.links = [...(parentGrp.links || []), grp];
        }
    });
    return root;
}

function getDummyAction(): NavItem[] {
    return [
        {
            id: '1',
            key: '1',
            name: 'Item ' + 1,
        },
        {
            id: '2',
            key: '2',
            name: 'Item ' + 2,
            isExpanded: true,
        },
        {
            id: '3',
            key: '3',
            name: 'Item ' + 3,
            isExpanded: true,
        },
        {
            id: '4',
            key: '4',
            name: 'Item ' + 4,
            visible: false,
        },
        {
            id: '5',
            key: '5',
            name: 'Item ' + 5,
            parentItemKey: '3',
        },
        {
            id: '6',
            key: '6',
            name: 'Item ' + 6,
            parentItemKey: '2',
            isExpanded: false,
        },
        {
            id: '7',
            key: '7',
            name: 'Item ' + 7,
            parentItemKey: '6',
        },
        {
            id: '8',
            key: '8',
            name: 'Item ' + 8,
            parentItemKey: '7',
            isExpanded: true,
        },
        {
            id: '9',
            key: '9',
            name: 'Item ' + 9,
            parentItemKey: '8',
        },
        {
            id: '10',
            key: '10',
            name: 'Item ' + 10,
            parentItemKey: '9',
            isExpanded: true,
        },
        {
            id: '11',
            key: '11',
            name: 'Item ' + 11,
        },
        {
            id: '12',
            key: '12',
            name: 'Item ' + 12,
        },
        {
            id: '13',
            key: '13',
            name: 'Item ' + 13,
            parentItemKey: '12',
        },
        {
            id: '14',
            key: '14',
            name: 'Item ' + 14,
            parentItemKey: '12',
        },
        {
            id: '15',
            key: '15',
            name: 'Item ' + 12,
            parentItemKey: '15',
        },
    ] as unknown as NavItem[];
}
