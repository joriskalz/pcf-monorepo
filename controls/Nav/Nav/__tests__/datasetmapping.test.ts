import { getMenuItemsFromDataset } from '../components/DatasetMapping';
import { ItemColumns } from '../ManifestConstants';
import { MockDataSet, MockEntityRecord } from '../__mocks__/mock-datasets';

describe('DatasetMapping', () => {
    beforeEach(() => jest.clearAllMocks());

    it('returns correct props', () => {
        const items = [
            new MockEntityRecord('1', {
                [ItemColumns.Key]: 'commandOpen',
                [ItemColumns.DisplayName]: 'Open',
                [ItemColumns.IconName]: 'Open',
                [ItemColumns.IconColor]: 'blue',
                [ItemColumns.Enabled]: true,
            }),
            new MockEntityRecord('2', {
                [ItemColumns.Key]: 'commandNew',
                [ItemColumns.DisplayName]: 'New',
                [ItemColumns.IconName]: 'New',
                [ItemColumns.IconColor]: 'blue',
                [ItemColumns.Enabled]: false,
            }),
            new MockEntityRecord('3', {
                [ItemColumns.Key]: 'commandSave',
                [ItemColumns.DisplayName]: 'Save',
                [ItemColumns.IconName]: 'Save',
                [ItemColumns.IconColor]: 'green',
                [ItemColumns.Enabled]: false,
            }),
            // Sub Items First Level
            new MockEntityRecord('5', {
                [ItemColumns.Key]: 'commandNewFrom',
                [ItemColumns.DisplayName]: 'NewFrom',
                [ItemColumns.IconName]: 'New',
                [ItemColumns.IconColor]: 'blue',
                [ItemColumns.Enabled]: false,
                [ItemColumns.ParentKey]: 'commandNew',
            }),
            new MockEntityRecord('6', {
                [ItemColumns.Key]: 'commandSaveAndClose',
                [ItemColumns.DisplayName]: 'SaveAndClose',
                [ItemColumns.IconName]: 'Save',
                [ItemColumns.IconColor]: 'green',
                [ItemColumns.Enabled]: false,
                [ItemColumns.ParentKey]: 'commandSave',
            }),
            // Sub Items Second Level
            new MockEntityRecord('7', {
                [ItemColumns.Key]: 'commandNewFromSub1',
                [ItemColumns.DisplayName]: 'NewFromSub1',
                [ItemColumns.IconName]: 'New',
                [ItemColumns.IconColor]: 'blue',
                [ItemColumns.Enabled]: false,
                [ItemColumns.ParentKey]: 'commandNewFrom',
            }),
            new MockEntityRecord('8', {
                [ItemColumns.Key]: 'commandNewFromSub2',
                [ItemColumns.DisplayName]: 'NewFromSub2',
                [ItemColumns.IconName]: 'New',
                [ItemColumns.IconColor]: 'blue',
                [ItemColumns.Enabled]: false,
                [ItemColumns.ParentKey]: 'commandNewFrom',
            }),
        ];

        const actions = getMenuItemsFromDataset(new MockDataSet(items));
        expect(actions).toMatchSnapshot();
    });
});
