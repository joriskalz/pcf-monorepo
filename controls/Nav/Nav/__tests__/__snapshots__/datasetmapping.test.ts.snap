// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetMapping returns correct props 1`] = `
Array [
  Object {
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "1",
      "values": Object {
        "ItemDisplayName": "Open",
        "ItemEnabled": true,
        "ItemIconColor": "blue",
        "ItemIconName": "Open",
        "ItemKey": "commandOpen",
      },
    },
    "enabled": true,
    "iconColor": "blue",
    "iconName": "Open",
    "id": "1",
    "isExpanded": undefined,
    "key": "commandOpen",
    "name": "Open",
    "parentItemKey": undefined,
    "visible": undefined,
  },
  Object {
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "2",
      "values": Object {
        "ItemDisplayName": "New",
        "ItemEnabled": false,
        "ItemIconColor": "blue",
        "ItemIconName": "New",
        "ItemKey": "commandNew",
      },
    },
    "enabled": false,
    "iconColor": "blue",
    "iconName": "New",
    "id": "2",
    "isExpanded": undefined,
    "key": "commandNew",
    "name": "New",
    "parentItemKey": undefined,
    "visible": undefined,
  },
  Object {
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "3",
      "values": Object {
        "ItemDisplayName": "Save",
        "ItemEnabled": false,
        "ItemIconColor": "green",
        "ItemIconName": "Save",
        "ItemKey": "commandSave",
      },
    },
    "enabled": false,
    "iconColor": "green",
    "iconName": "Save",
    "id": "3",
    "isExpanded": undefined,
    "key": "commandSave",
    "name": "Save",
    "parentItemKey": undefined,
    "visible": undefined,
  },
  Object {
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "5",
      "values": Object {
        "ItemDisplayName": "NewFrom",
        "ItemEnabled": false,
        "ItemIconColor": "blue",
        "ItemIconName": "New",
        "ItemKey": "commandNewFrom",
        "ItemParentKey": "commandNew",
      },
    },
    "enabled": false,
    "iconColor": "blue",
    "iconName": "New",
    "id": "5",
    "isExpanded": undefined,
    "key": "commandNewFrom",
    "name": "NewFrom",
    "parentItemKey": "commandNew",
    "visible": undefined,
  },
  Object {
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "6",
      "values": Object {
        "ItemDisplayName": "SaveAndClose",
        "ItemEnabled": false,
        "ItemIconColor": "green",
        "ItemIconName": "Save",
        "ItemKey": "commandSaveAndClose",
        "ItemParentKey": "commandSave",
      },
    },
    "enabled": false,
    "iconColor": "green",
    "iconName": "Save",
    "id": "6",
    "isExpanded": undefined,
    "key": "commandSaveAndClose",
    "name": "SaveAndClose",
    "parentItemKey": "commandSave",
    "visible": undefined,
  },
  Object {
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "7",
      "values": Object {
        "ItemDisplayName": "NewFromSub1",
        "ItemEnabled": false,
        "ItemIconColor": "blue",
        "ItemIconName": "New",
        "ItemKey": "commandNewFromSub1",
        "ItemParentKey": "commandNewFrom",
      },
    },
    "enabled": false,
    "iconColor": "blue",
    "iconName": "New",
    "id": "7",
    "isExpanded": undefined,
    "key": "commandNewFromSub1",
    "name": "NewFromSub1",
    "parentItemKey": "commandNewFrom",
    "visible": undefined,
  },
  Object {
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "8",
      "values": Object {
        "ItemDisplayName": "NewFromSub2",
        "ItemEnabled": false,
        "ItemIconColor": "blue",
        "ItemIconName": "New",
        "ItemKey": "commandNewFromSub2",
        "ItemParentKey": "commandNewFrom",
      },
    },
    "enabled": false,
    "iconColor": "blue",
    "iconName": "New",
    "id": "8",
    "isExpanded": undefined,
    "key": "commandNewFromSub2",
    "name": "NewFromSub2",
    "parentItemKey": "commandNewFrom",
    "visible": undefined,
  },
]
`;
